src/genparticles.cxx

```diff
@@ -0,0 +1,305 @@
+    IS_ELE_FROM_TAU = 3,
+    IS_MUON_FROM_TAU = 4,
+    IS_TAU_HAD_DECAY = 5,
+    IS_FAKE = 6
+};
+
+namespace genmatching {
+
+namespace tau {
+/**
+ * @brief implementation of the genmatching used in tau analyses, based
+ * on
+ * https://github.com/KIT-CMS/Artus/blob/dictchanges/KappaAnalysis/src/Utility/GeneratorInfo.cc
+ * implementation. The genmatches are represented as integer flags:
+    Decaytype         | Value
+    ------------------|-------
+    NONE              | 1,
+    IS_ELE_PROMPT     | 1,
+    IS_MUON_PROMPT    | 2,
+    IS_ELE_FROM_TAU   | 3,
+    IS_MUON_FROM_TAU  | 4,
+    IS_TAU_HAD_DECAY  | 5,
+    IS_FAKE           | 6
+
+    The genmatch is caluclated individually for each taupair candidate.
+
+ *
+ * @param df The dataframe to be used for the genmatching.
+ * @param outputname The name of the output column.
+ * @param hadronicGenTaus The name of the column containing the hadronicGenTaus
+ from genmatching::tau::hadronicGenTaus
+ * @param genparticles_pdgid The name of the column containing the pdgids of the
+ genparticles
+ * @param genparticles_statusFlag The name of the column containing the
+ statusFlags of the genparticles
+ * @param genparticles_pt The name of the column containing the pT of the
+ genparticles
+ * @param genparticles_eta The name of the column containing the eta of the
+ genparticles
+ * @param genparticles_phi The name of the column containing the phi of the
+ genparticles
+ * @param genparticles_mass The name of the column containing the mass of the
+ genparticles
+ * @param lepton_p4 The name of the column containing the p4 of the lepton to be
+ genmatched
+ * @return a dataframe with the genmatching as a column named outputname
+ */
+ROOT::RDF::RNode genmatching(ROOT::RDF::RNode df, const std::string &outputname,
+                             const std::string &hadronicGenTaus,
+                             const std::string &genparticles_pdgid,
+                             const std::string &genparticles_statusFlag,
+                             const std::string &genparticles_pt,
+                             const std::string &genparticles_eta,
+                             const std::string &genparticles_phi,
+                             const std::string &genparticles_mass,
+                             const std::string &lepton_p4) {
+    auto match_lepton = [](const std::vector<int> &hadronicGenTaus,
+                           const ROOT::RVec<int> &pdgids,
+                           const ROOT::RVec<int> &status_flags,
+                           const ROOT::RVec<float> &pts,
+                           const ROOT::RVec<float> &etas,
+                           const ROOT::RVec<float> &phis,
+                           const ROOT::RVec<float> &masses,
+                           const ROOT::Math::PtEtaPhiMVector &lepton_p4) {
+        // find closest lepton fulfilling the requirements
+        float min_delta_r = 9999;
+        int closest_genparticle_index = 0;
```

Comment:
In the code it is taken care of by also looking at the `min_delta_r`, but generally you would find a particle at index 0. I would suggest to change that to -1 or something like that to make clear that no particle was found.